vm/src/slots.rs

```diff
@@ -148,10 +163,173 @@ pub trait SlotDescriptor: PyValue {
 pub trait Hashable: PyValue {
     #[pyslot]
     fn tp_hash(zelf: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyHash> {
-        let zelf = <PyRef<Self>>::try_from_object(vm, zelf)?;
+        let zelf = PyRef::try_from_object(vm, zelf)?;
         Self::hash(zelf, vm)
     }
 
     #[pymethod(magic)]
     fn hash(zelf: PyRef<Self>, vm: &VirtualMachine) -> PyResult<PyHash>;
 }
+
+#[pyimpl]
+pub trait Comparable: PyValue {
+    #[pyslot]
+    fn tp_cmp(
+        zelf: PyObjectRef,
+        other: PyObjectRef,
+        op: PyComparisonOp,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        let zelf = PyRef::try_from_object(vm, zelf)?;
+        Self::cmp(zelf, other, op, vm)
+    }
+
+    fn cmp(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        op: PyComparisonOp,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue>;
+
+    #[pymethod(magic)]
+    fn eq(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        Self::cmp(zelf, other, PyComparisonOp::Eq, vm)
+    }
+    #[pymethod(magic)]
+    fn ne(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        Self::cmp(zelf, other, PyComparisonOp::Ne, vm)
+    }
+    #[pymethod(magic)]
+    fn lt(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        Self::cmp(zelf, other, PyComparisonOp::Lt, vm)
+    }
+    #[pymethod(magic)]
+    fn le(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        Self::cmp(zelf, other, PyComparisonOp::Le, vm)
+    }
+    #[pymethod(magic)]
+    fn ge(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        Self::cmp(zelf, other, PyComparisonOp::Ge, vm)
+    }
+    #[pymethod(magic)]
+    fn gt(
+        zelf: PyRef<Self>,
+        other: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        Self::cmp(zelf, other, PyComparisonOp::Gt, vm)
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum PyComparisonOp {
+    Lt,
+    Le,
+    Eq,
+    Ne,
+    Ge,
+    Gt,
+}
```

Comment:
is this mandatory for python spec or a utillity to allow tp_cmp? I thought this is CPython internal stuff.
If we can avoid enum, i prefer that way. We need to think more about inlining if we use enum for that.
(I didn't check it is possible - but maybe we can have 6 tp_ slots for comparison?)

Otherwise, appending `#[inline]` for `cmp` maybe helpful to inline it when we call them with constant enum for each operation for specific type not through tp_cmp.

Reply:
I think this enum actually makes code a lot more readable; we were already doing a lot of

```rust
fn cmp() {}
fn lt() { self.cmp(|a, b| a < b) }
```

So this just makes it more standardized and less repetitive, rather than 10 types with 6 different comparison op functions that all call `self.cmp(|a, b| a OP b)`, we can deduplicate a lot of it and pass data around (`PyComparisonOp`) instead of functionality (`F: Fn(usize, usize) -> bool`). And yeah, I think it probably would be good to try `#[inline]`-ing the `cmp()` function to avoid branching in a loop, if possible.